tensorflow_graphics/projects/mesh_rcnn/ops/cubify.py

```diff
@@ -0,0 +1,262 @@
+# Copyright 2020 The TensorFlow Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Implementation of the cubify operation for Mesh R-CNN."""
+
+import tensorflow as tf
+
+from tensorflow_graphics.projects.mesh_rcnn.structures.mesh import Meshes
+from tensorflow_graphics.util import shape
+
+
+def _ravel_index(index, dims):
+  """
+  Computes a linear index in an array of shape dims.
+  Reverse functionality of tf.unravel_index.
+  Args:
+    index: A Tensor with each row corresponding to indices into an array
+    of dimension dims.
+    dims: The shape of the array to be indexed.
+
+  Note:
+
+    The shorthands used below and in the code are
+      `H`: height of input space
+      `W`: width of input space
+      `D`: depth of input space
+
+    Only supports dims=(H,W,D)
+
+  Returns:
+    Integer Tensor containing the index into an array of shape `dims`.
+  """
+  shape.check_static(tensor=index, has_rank=2)
+  if len(dims) != 3:
+    raise ValueError('Expects a 3-element list')
+  if index.shape[1] != 3:
+    raise ValueError('Expects an index tensor of shape Nx3')
+  _, W, D = dims  # pylint: disable=invalid-name
+  linear_index = index[:, 0] * W * D + index[:, 1] * D + index[:, 2]
+  return linear_index
+
+
+def cubify(voxel_grid, threshold=0.5):
+  """
+  Converts voxel occupancy probability grids into a triangle mesh.
+
+  Each occupied voxel is replaced
+  with a cuboid triangle mesh with 8 vertices, 18 edges, and
+  12 faces. Shared vertices and edges between adjacent occupied voxels are
+  merged, and shared interior faces are eliminated. This results in a
+  watertight mesh whose topology depends on the voxel occupancy probabilities.
+
+  The shorthands used below and in the code are
+    `V`: The number of vertices.
+    `C`: The number of channels in the input data.
+    `N`: Batch dimension
+    `D`: depth of input space (representing z coordinates)
+    `W`: width of input space (representing x coordinates)
+    `H`: height of input space (representing y coordinates)
+
+    The coordinates assume a Y-up convention.
+
+  Args:
+    voxel_grid: flaot32 tensor of shape `[N, D, H, W]` containing the voxel
```

Comment:
A `float32` tensor of shape ...
Can this generalize to arbitary batch dimensions?
`[A1, A2, .., N, D, H, W]`